Libplanet/Blocks/BlockMetadata.cs

```diff
@@ -0,0 +1,301 @@
+        public DateTimeOffset Timestamp
+        {
+            get => _timestamp;
+            set => _timestamp = value.ToUniversalTime();
+        }
+
+        /// <inheritdoc cref="IBlockMetadata.Miner"/>
+        public Address Miner { get; set; }
+
+        /// <inheritdoc cref="IBlockMetadata.Difficulty"/>
+        /// <exception cref="InvalidBlockDifficultyException">Thrown when the value to set is
+        /// negative.</exception>
+        /// <remarks>This cannot not be negative.
+        /// <para>When <see cref="Difficulty"/> is updated, <see cref="TotalDifficulty"/> is also
+        /// updated together.  For example, when <see cref="Difficulty"/> = 10 and
+        /// <see cref="TotalDifficulty"/> = 50, if <see cref="Difficulty"/> is updated to
+        /// 20 (= 10 + 10) <see cref="TotalDifficulty"/> is also updated to 60 (= 50 + 10).</para>
+        /// </remarks>
+        public long Difficulty
+        {
+            get => _difficulty;
+            set
+            {
+                if (value < 0L)
+                {
+                    throw new InvalidBlockDifficultyException(
+                        $"{nameof(Difficulty)} cannot be negative: {value}"
+                    );
+                }
+
+                long delta = value - _difficulty;
+                _difficulty = value;
+                _totalDifficulty += delta;
+            }
+        }
+
+        /// <inheritdoc cref="IBlockMetadata.TotalDifficulty"/>
+        /// <exception cref="InvalidBlockTotalDifficultyException">Thrown when the value to set
+        /// is less than <see cref="Difficulty"/>.</exception>
+        public BigInteger TotalDifficulty
+        {
+            get => _totalDifficulty;
+            set
+            {
+                if (value < BigInteger.Zero)
+                {
+                    throw new InvalidBlockTotalDifficultyException(
+                        Difficulty,
+                        value,
+                        $"{nameof(TotalDifficulty)} cannot be negative: ${value}."
+                    );
+                }
+                else if (value < Difficulty)
+                {
+                    string msg =
+                        $"{nameof(TotalDifficulty)} ({value}) cannot be less than " +
+                        $"{nameof(Difficulty)} ({Difficulty}).";
+                    throw new InvalidBlockTotalDifficultyException(Difficulty, value, msg);
+                }
+
+                _totalDifficulty = value;
+            }
+        }
+
+        /// <inheritdoc cref="IBlockMetadata.PreviousHash"/>
+        public BlockHash? PreviousHash { get; set; }
+
+        /// <inheritdoc cref="IBlockMetadata.TxHash"/>
+        [SuppressMessage(
+            "SonarQube",
+            "S2292",
+            Justification = "The backing field purposes to prevent intercepting from subclasses.")]
+        public virtual HashDigest<SHA256>? TxHash
+        {
+            get => _txHash;
+            set => _txHash = value;
```

Comment:
```suggestion
            protected set => _txHash = value;
```

Do we need `set` to be publicly exposed? The same goes for `BlockContent<T>`.